src/index/diskann/diskann.cc

```diff
@@ -376,48 +379,65 @@ DiskANNIndexNode<T>::Prepare(const Config& cfg) {
-                            << ") is larger than 1/3 of the total data number.";
-        return false;
-    }
-    if (num_nodes_to_cache > 0) {
-        std::vector<uint32_t> node_list;
-        LOG_KNOWHERE_INFO_ << "Caching " << num_nodes_to_cache << " sample nodes around medoid(s).";
-        if (prep_conf.use_bfs_cache) {
-            auto gen_cache_expect = TryDiskANNCall<bool>([&]() -> bool {
-                pq_flash_index_->cache_bfs_levels(num_nodes_to_cache, node_list);
-                return true;
-            });
-
-            if (!gen_cache_expect.has_value()) {
-                LOG_KNOWHERE_ERROR_ << "Failed to generate bfs cache for DiskANN.";
-                return false;
-            }
-
+    auto cached_nodes_file = diskann::get_cached_nodes_file(index_prefix_);
+    if (file_exists(cached_nodes_file)) {
+        LOG_KNOWHERE_INFO_ << "Reading cached nodes from file.";
+        auto gen_cache_from_file_expect = TryDiskANNCall<bool>([&]() -> bool {
+            pq_flash_index_->load_cache_list_from_file(cached_nodes_file);
+            return true;
+        });
+        if (!gen_cache_from_file_expect.has_value()) {
+            LOG_KNOWHERE_ERROR_ << "Failed to load cache for DiskANN from cached nodes file.";
+            return false;
+        }
+    } else {
+        auto num_nodes_to_cache = GetCachedNodeNum(prep_conf.search_cache_budget_gb, pq_flash_index_->get_data_dim(),
+                                                   pq_flash_index_->get_max_degree());
+        if (num_nodes_to_cache > pq_flash_index_->get_num_points() / 3) {
+            LOG_KNOWHERE_ERROR_ << "Failed to generate cache, num_nodes_to_cache(" << num_nodes_to_cache
+                                << ") is larger than 1/3 of the total data number.";
+            return false;
+        }
+        if (num_nodes_to_cache > 0) {
+            LOG_KNOWHERE_INFO_ << "Caching " << num_nodes_to_cache << " sample nodes around medoid(s).";
         } else {
-            auto gen_cache_expect = TryDiskANNCall<bool>([&]() -> bool {
-                pq_flash_index_->generate_cache_list_from_sample_queries(warmup_query_file, 15, 6, num_nodes_to_cache,
-                                                                         node_list);
+            std::vector<uint32_t> node_list;
+            if (prep_conf.use_bfs_cache) {
+                LOG_KNOWHERE_INFO_ << "Use bfs to generate cache list";
+                auto gen_cache_expect = TryDiskANNCall<bool>([&]() -> bool {
+                    pq_flash_index_->cache_bfs_levels(num_nodes_to_cache, node_list);
+                    return true;
+                });
+
+                if (!gen_cache_expect.has_value()) {
+                    LOG_KNOWHERE_ERROR_ << "Failed to generate bfs cache for DiskANN.";
+                    return false;
+                }
+
+            } else {
+                LOG_KNOWHERE_INFO_ << "Use sample_queries to generate cache list";
```

Comment:
Still generate cache if `num_nodes_to_cache <= 0` and `prep_conf.use_bfs_cache = false`?

Reply:
Fixed